### ISSUE TITLE
torch.norm gives NaN gradient when I input small-value float16 tensor

### ISSUE THREAD

**Issue by KiaLAN** [pytorch/pytorch#43211]
## To Reproduce

```
import torch
import torch.nn as nn
import torch.nn.functional as F
torch.autograd.set_detect_anomaly(False)

a = torch.Tensor([1e-4]).cuda().half()
a.requires_grad=True
a.norm(3).backward()
a.grad
```

The result of `a.grad` is `tensor([nan], device='cuda:0', dtype=torch.float16)`.

I couldn't produce the behavior when using float32. I guess this is a float16 related bug.

## Expected behavior

I think the gradient should be 0.

## Environment

```
PyTorch version: 1.6.0
Is debug build: False
CUDA used to build PyTorch: 10.2

OS: Ubuntu 16.04.6 LTS (x86_64)
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609
Clang version: Could not collect
CMake version: version 3.5.1

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.2.89
GPU models and configuration: GPU 0: GeForce GTX 1050 Ti
Nvidia driver version: 440.33.01
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.11.0
[conda] _pytorch_select           0.2                       gpu_0    defaults
[conda] blas                      1.0                         mkl    defaults
[conda] cudatoolkit               10.0.130                      0    defaults
[conda] mkl                       2020.0                      166    defaults
[conda] mkl-service               2.3.0            py37he904b0f_0    defaults
[conda] mkl_fft                   1.0.15           py37ha843d7b_0    defaults
[conda] mkl_random                1.1.0            py37hd6b4f25_0    defaults
[conda] numpy                     1.18.1           py37h4f9e942_0    defaults
[conda] numpy-base                1.18.1           py37hde5b4d6_1    defaults
[conda] numpydoc                  0.9.2                      py_0    defaults
[conda] pytorch-gpu               1.3.1                         0    defaults
[conda] torch                     1.6.0                    pypi_0    pypi
[conda] torchvision               0.7.0                    pypi_0    pypi
```

## Additional context

cc @ezyang @SsnL @albanD @zou3519 @gqchen

**Comment by 07Agarg** [comment]
> Hi,
> 
> I guess this is just a number that is small enough that it generate inf for the derivative of sqrt but is not 0 and so our special case at 0 does not apply.
> I'm sure you can do similar things with float32 for very small numbers that are not 0.
> 
> A simple temporary fix I would recommend is to clamp small values that are very close to the numerical precision of float16 to be exactly `0`.

Hi Alban,
Can you please give a sample code to resolve this issue?

Thanks.

**Comment by albanD** [comment]
Hi,

Sure the following modified code gives the expected result:
```python
import torch
import torch.nn as nn
import torch.nn.functional as F
torch.autograd.set_detect_anomaly(False)

a = torch.Tensor([1e-4]).cuda().half()
a.requires_grad=True

# Clamp to avoid values that are too low
# Disable autograd while clamping because
# we only do this for numerical stability
with torch.no_grad():
    a.clamp_(min=1e-2)

a.norm(3).backward()
a.grad
```

**Comment by 07Agarg** [comment]
> Hi,
> 
> Sure the following modified code gives the expected result:
> 
> ```python
> import torch
> import torch.nn as nn
> import torch.nn.functional as F
> torch.autograd.set_detect_anomaly(False)
> 
> a = torch.Tensor([1e-4]).cuda().half()
> a.requires_grad=True
> 
> # Clamp to avoid values that are too low
> # Disable autograd while clamping because
> # we only do this for numerical stability
> with torch.no_grad():
>     a.clamp_(min=1e-2)
> 
> a.norm(3).backward()
> a.grad
> ```

Thanks Alban!
